pkg/intervals/intervals.go

```diff
@@ -0,0 +1,281 @@
+package intervals
+
+import (
+	"cmp"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+	"unicode/utf8"
+)
+
+type (
+	// Basis contains constants and functions for work with values of type T
+	Basis[T any] struct {
+		Min  T                  // min value of the type T
+		Max  T                  // max value of the type T
+		CmpF func(v1, v2 T) int // comparison function of T type values
+	}
+
+	// Interval represents an interval of values of type T,
+	// interval can be open, half-open, closed.
+	Interval[T any] struct {
+		L   T
+		R   T
+		B   Basis[T]
```

Comment:
What is the reason that every interval object has the basis? Why not to build a set of intervals for the same basis, so the basis is the parameter of the Intervals builder?

Reply:
Basis now produces the intervals and contains all the operations to handle intersection, union, etc.